src/sources/exec.rs

```diff
@@ -0,0 +1,830 @@
+    log_event.insert(log_schema().source_type_key(), Bytes::from(EXEC));
+
+    // Add data stream of stdin or stderr (if needed)
+    if let Some(data_stream) = data_stream {
+        log_event.insert(DATA_STREAM_KEY, data_stream.clone());
+    }
+
+    // Add pid (if needed)
+    if let Some(pid) = pid {
+        log_event.insert(PID_KEY, pid as i64);
+    }
+
+    // Add exit status (if needed)
+    if let Some(exit_status) = exit_status {
+        log_event.insert(EXIT_STATUS_KEY, exit_status as i64);
+    }
+
+    // Add exec duration millis (if needed)
+    if let Some(exec_duration_millis) = exec_duration_millis {
+        log_event.insert(EXEC_DURATION_MILLIS_KEY, *exec_duration_millis as i64);
+    }
+
+    // Add hostname (if needed)
+    if let Some(hostname) = hostname {
+        log_event.insert(log_schema().host_key(), hostname.clone());
+    }
+
+    // Add command
+    log_event.insert(COMMAND_KEY, config.command.clone());
+
+    Event::Log(log_event)
+}
+
+fn spawn_reader_thread<R: 'static + AsyncRead + Unpin + std::marker::Send>(
+    reader: BufReader<R>,
+    shutdown: ShutdownSignal,
+    event_per_line: bool,
+    buf_size: u64,
+    stream: &'static str,
+    sender: Sender<(Bytes, &'static str)>,
+) {
+    // Start the green background thread for collecting
+    Box::pin(tokio::spawn(async move {
+        debug!("Start capturing {} command output.", stream);
+
+        let mut read_buffer: Vec<u8> = Vec::new();
+
+        let reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
+
+        let mut limited_reader = reader.take(buf_size);
+
+        if event_per_line {
+            // Keep reading lines (lines longer than the max buffer will be split)
+            while let Ok(bytes_read) = limited_reader.read_until(b'\n', &mut read_buffer).await {
+                if bytes_read == 0 {
+                    // If we get a continuous stream of \n the bytes_read will be at least 1
+                    debug!("End of input reached, stop reading.");
+                    break;
+                }
+
+                // Strip of the end of line bytes
+                if read_buffer.ends_with(&[b'\n']) {
+                    let _ = read_buffer.pop();
+
+                    if read_buffer.ends_with(&[b'\r']) {
+                        let _ = read_buffer.pop();
+                    }
+                }
+
+                let read_bytes = Bytes::from(read_buffer.clone());
+                if sender.send((read_bytes, stream)).await.is_err() {
+                    break;
+                }
+
+                // Clear the read buffer ready for the next read
+                read_buffer.clear();
+
+                // Reset the limit for the next read (minus any left over if we do utf8 checks)
+                limited_reader.set_limit(buf_size);
+            }
+        } else {
+            // Keep reading max buffer chunks
+            while let Ok(bytes_read) = limited_reader.read_to_end(&mut read_buffer).await {
```

Comment:
I think I see what you are doing here. I would have actually expected the buffer size to be a hard limit on the output for `event_per_line = false`. What would you think of just reading up until `buf_size` and just dropping the rest?

Reply:
I think because the logic is currently applying event_per_line to both scheduled and streaming then it should not be a hard limit but just a limit for each read/event. If you don't want to allow event_per_line for streaming then we could make it a hard limit. I kind of feel that the main purpose of the maximum_buffer_size_bytes is so we don't blow up memory on read which the limit per read/event should do (so we don't need to make it a hard limit for no more data). Actually even if event_per_line only applied to scheduled we might get a streaming process that malfunctions and starts spewing out something without a newline and I don't think we would want to drop the rest of its output. Let's discuss further if needed.

Reply:
Changed it to use the codec now which does throw away lines that are too long (I'm not sure I like that behavior but I figured it was ok and we could always replace the codec later if needed).